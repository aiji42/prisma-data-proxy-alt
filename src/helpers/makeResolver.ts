import { PrismaClient } from "@prisma/client/scripts/default-index";
import { DMMF } from "@prisma/client/runtime";
import { SelectionNode } from "graphql";
import { IFieldResolver } from "@graphql-tools/utils";

const toLowerFirstLetter = (str: string) => {
  return str.charAt(0).toLowerCase() + str.substring(1);
};

const functionsMappings = (
  dmmf: DMMF.Document
): Record<string, { operation: string; model: string }> => {
  return Object.fromEntries(
    dmmf.mappings.modelOperations.flatMap(
      ({ model, plural, ...operations }) => {
        return Object.entries(operations).map(([k, v]) => [
          v,
          { operation: k, model },
        ]);
      }
    )
  );
};

interface AggregateField {
  [x: string]: true | AggregateField;
}

const digAggregateField = (
  selections: readonly SelectionNode[],
  isRoot: boolean = true
): AggregateField => {
  return selections.reduce((res, selection) => {
    if (
      "name" in selection &&
      "selectionSet" in selection &&
      !(isRoot && !selection.name.value.startsWith("_"))
    ) {
      const dug = digAggregateField(
        selection.selectionSet?.selections ?? [],
        false
      );
      return {
        ...res,
        [selection.name.value]: Object.keys(dug).length ? dug : true,
      };
    }
    return res;
  }, {});
};

export function convertResultToPrismaFormat(obj: any) {
  if (typeof obj === "bigint") {
    return {
      prisma__type: "bigint",
      prisma__value: String(obj),
    };
  }
  if (typeof obj === "number") {
    return {
      prisma__type: "number",
      prisma__value: obj,
    };
  }
  if (typeof obj === "string") {
    return {
      prisma__type: "string",
      prisma__value: obj,
    };
  }
  if (obj == null) {
    return {
      prisma__type: "null",
      prisma__value: null,
    };
  }
  if (typeof obj === "boolean") {
    return {
      prisma__type: "bool",
      prisma__value: obj,
    };
  }
  if (obj instanceof Date) {
    return {
      prisma__type: "datetime",
      prisma__value: obj,
    };
  }
  if (typeof obj === "object") {
    for (const key in obj) {
      obj[key] = convertResultToPrismaFormat(obj[key]);
    }
  }
  return obj;
}

export const rootOperationProxy = (db: PrismaClient, dmmf: DMMF.Document) => {
  const mapping = functionsMappings(dmmf);
  return new Proxy(
    {},
    {
      get: (_, method): IFieldResolver<Record<string, unknown>, unknown> => {
        return async (...[, args, , info]) => {
          if (method.toString() === "queryRaw") {
            let result = await db.$queryRawUnsafe(
              args.query,
              ...JSON.parse(args.parameters)
            );
            return convertResultToPrismaFormat(result);
          }

          const { operation, model } = mapping[method.toString()];

          const newArgs =
            operation === "aggregate" || operation === "groupBy"
              ? {
                  ...digAggregateField(
                    info.fieldNodes[0].selectionSet?.selections ?? []
                  ),
                  ...args,
                }
              : args;

          return db[toLowerFirstLetter(model)][operation.replace(/One$/, "")](
            newArgs
          );
        };
      },
    }
  );
};

export const relatedOperationProxy = (
  db: PrismaClient,
  { name, fields, primaryKey }: DMMF.Model
) => {
  return new Proxy(
    {},
    {
      get: (
        _,
        method
      ): IFieldResolver<Record<string, unknown>, unknown> | undefined => {
        const field = fields.find(({ name }) => name === method.toString());
        if (!field?.relationName) return undefined;

        const idFieldNames: string[] = fields
          .filter((field) => {
            return field.isId || primaryKey?.fields?.includes(field.name);
          })
          .map((x) => x.name);

        if (idFieldNames.length == 0) return undefined;

        return async (parent, args) => {
          let where: Record<string, any>;

          if (idFieldNames.length > 1) {
            const conditions: Record<string, any> = {};
            for (const idFieldName of idFieldNames) {
              conditions[idFieldName] = parent[idFieldName];
            }
            where = {
              [idFieldNames.join("_")]: conditions,
            };
          } else {
            where = {
              [idFieldNames[0]]: parent[idFieldNames[0]],
            };
          }

          return await db[toLowerFirstLetter(name)]
            .findUnique({ where })
            [method](args);
        };
      },
    }
  );
};

export const makeResolver = (dmmf: DMMF.Document, db: PrismaClient) => {
  return {
    ...Object.fromEntries(
      dmmf.datamodel.models.map((model) => [
        model.name,
        relatedOperationProxy(db, model),
      ])
    ),
    Query: rootOperationProxy(db, dmmf),
    Mutation: rootOperationProxy(db, dmmf),
  };
};
