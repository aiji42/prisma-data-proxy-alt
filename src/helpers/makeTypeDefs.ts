import { gql } from "apollo-server-express";
import { DMMF } from "@prisma/client/runtime";

export const makeTypeDefs = (dmmf: DMMF.Document) => {
  const enums = [
    ...dmmf.schema.enumTypes.prisma,
    ...(dmmf.schema.enumTypes.model ?? []),
  ].map(({ name, values }) => {
    return `
      enum ${name} {
        ${values.join(" ")}
      }
    `;
  });

  const types = [
    ...dmmf.schema.outputObjectTypes.prisma,
    ...dmmf.schema.outputObjectTypes.model,
  ].map((m) => {
    return `
      type ${m.name} {
        ${m.fields
          .map((f) => {
            return `${f.name}${
              f.args.length > 0
                ? `(${f.args.map(({ name }) => `${name}: Any`).join(" ")})`
                : ""
            }: ${
              f.outputType.isList
                ? `[${f.outputType.type}]`
                : `${f.outputType.type}`
            }${f.isNullable ? "" : "!"}`;
          })
          .join(" ")}
      }
      `;
  });

  return gql`
    scalar Any
    scalar DateTime
    scalar Json
    scalar BigInt
    ${enums}
    ${types}
  `;
};
