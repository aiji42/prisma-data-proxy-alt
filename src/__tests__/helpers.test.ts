import { describe, expect, test, vi } from "vitest";
import { getSampleDMMF } from "./__fixtures__/getSampleSchema";
import { makeTypeDefs } from "../helpers/makeTypeDefs";
import {
  rootOperationProxy,
  relatedOperationProxy,
  convertResultToPrismaFormat,
} from "../helpers/makeResolver";
import { parse } from "graphql";

describe("makeTypeDefs", () => {
  test("Check the generated graphql schema.", async () => {
    const dmmf = await getSampleDMMF();
    const typeDef = makeTypeDefs(dmmf);

    expect(typeDef.loc?.source.body).toMatchSnapshot();
  });
});

describe("convertResultToPrismaFormat", () => {
  test("Convert query raw result to prisma format", async () => {
    const result = {
      id: 1,
      name: "test",
      created_at: new Date(2000, 1, 1),
      deleted: false,
      count: 100n,
      team: {
        id: 10,
      },
      country: null,
    };
    const converted = convertResultToPrismaFormat(result);

    expect(converted).toEqual({
      id: {
        prisma__type: "number",
        prisma__value: 1,
      },
      name: {
        prisma__type: "string",
        prisma__value: "test",
      },
      created_at: {
        prisma__type: "datetime",
        prisma__value: new Date(2000, 1, 1),
      },

      deleted: {
        prisma__type: "bool",
        prisma__value: false,
      },
      count: {
        prisma__type: "bigint",
        prisma__value: "100",
      },
      team: {
        id: {
          prisma__type: "number",
          prisma__value: 10,
        },
      },
      country: {
        prisma__type: "null",
        prisma__value: null,
      },
    });
  });
});

describe("makeResolver > rootOperationProxy", () => {
  test("called fineManyUser", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          findMany: mock,
        },
      },
      dmmf
    );

    // @ts-ignore
    proxy.findManyUser({}, { foo: "foo", bar: "bar" });
    expect(mock).toBeCalledWith({ foo: "foo", bar: "bar" });
  });

  test("called createOne", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          create: mock,
        },
      },
      dmmf
    );

    // @ts-ignore
    proxy.createOneUser(
      {},
      {
        data: {
          name: "foo",
        },
      }
    );
    expect(mock).toBeCalledWith({
      data: {
        name: "foo",
      },
    });
  });

  test("called aggregateUser", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          aggregate: mock,
        },
      },
      dmmf
    );

    const a = parse(`query {
      aggregateUser(where: {
        id: 1
      }) {
        _count {
          _all
          id
        }
      }
    }`);

    // @ts-ignore
    proxy.aggregateUser(
      {},
      {
        where: {
          id: 1,
        },
      },
      null,
      {
        // @ts-ignore
        fieldNodes: a.definitions[0].selectionSet.selections,
      }
    );
    expect(mock).toBeCalledWith({
      where: {
        id: 1,
      },
      _count: { _all: true, id: true },
    });
  });

  test("called groupByUser", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const proxy = rootOperationProxy(
      {
        user: {
          groupBy: mock,
        },
      },
      dmmf
    );

    const a = parse(`query {
      groupByUser(by: [name]) {
        _max {
          id
        }
        name
      }
    }`);

    // @ts-ignore
    proxy.groupByUser({}, { by: ["name"] }, null, {
      // @ts-ignore
      fieldNodes: a.definitions[0].selectionSet.selections,
    });
    expect(mock).toBeCalledWith({
      by: ["name"],
      _max: { id: true },
    });
  });
});

describe("makeResolver > relatedOperationProxy", () => {
  test("called user.Team", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const relatedMock = vi.fn();
    const proxy = relatedOperationProxy(
      {
        user: {
          findUnique: mock.mockReturnValue({ Team: relatedMock }),
        },
      },
      dmmf.datamodel.models.find(({ name }) => name === "User")!
    );

    // @ts-ignore
    proxy.Team({ id: 10, name: "foo" }, { select: { id: true, name: true } });
    expect(mock).toBeCalledWith({ where: { id: 10 } });
    expect(relatedMock).toBeCalledWith({ select: { id: true, name: true } });
  });

  test("called leaderboardRow.User", async () => {
    const dmmf = await getSampleDMMF();
    const mock = vi.fn();
    const relatedMock = vi.fn();
    const proxy = relatedOperationProxy(
      {
        leaderboardRow: {
          findUnique: mock.mockReturnValue({ User: relatedMock }),
        },
      },
      dmmf.datamodel.models.find(({ name }) => name === "LeaderboardRow")!
    );

    // @ts-ignore
    proxy.User(
      { leaderboardId: 1, userId: 10, rating: 1200 },
      { select: { id: true, name: true } }
    );
    expect(mock).toBeCalledWith({
      where: { leaderboardId_userId: { leaderboardId: 1, userId: 10 } },
    });
    expect(relatedMock).toBeCalledWith({ select: { id: true, name: true } });
  });
});
