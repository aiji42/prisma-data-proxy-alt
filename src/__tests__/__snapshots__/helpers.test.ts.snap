// Vitest Snapshot v1

exports[`makeTypeDefs > Check the generated graphql schema. 1`] = `
"
    scalar Any
    scalar DateTime
    scalar Json
    scalar BigInt
    scalar Decimal
    
      enum UserScalarFieldEnum {
        id email name teamId createdAt updatedAt
      }
    ,
      enum TeamScalarFieldEnum {
        id name language createdAt updatedAt
      }
    ,
      enum LeaderboardRowScalarFieldEnum {
        leaderboardId userId rating
      }
    ,
      enum SortOrder {
        asc desc
      }
    ,
      enum QueryMode {
        default insensitive
      }
    ,
      enum Language {
        JAPANESE ENGLISH CHINESE FRENCH GERMAN SPANISH
      }
    
    
      type Query {
        findFirstUser(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): User findManyUser(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [User]! aggregateUser(where: Any orderBy: Any cursor: Any take: Any skip: Any): AggregateUser! groupByUser(where: Any orderBy: Any by: Any having: Any take: Any skip: Any): [UserGroupByOutputType]! findUniqueUser(where: Any): User findFirstTeam(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): Team findManyTeam(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [Team]! aggregateTeam(where: Any orderBy: Any cursor: Any take: Any skip: Any): AggregateTeam! groupByTeam(where: Any orderBy: Any by: Any having: Any take: Any skip: Any): [TeamGroupByOutputType]! findUniqueTeam(where: Any): Team findFirstLeaderboardRow(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): LeaderboardRow findManyLeaderboardRow(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [LeaderboardRow]! aggregateLeaderboardRow(where: Any orderBy: Any cursor: Any take: Any skip: Any): AggregateLeaderboardRow! groupByLeaderboardRow(where: Any orderBy: Any by: Any having: Any take: Any skip: Any): [LeaderboardRowGroupByOutputType]! findUniqueLeaderboardRow(where: Any): LeaderboardRow
      }
      ,
      type Mutation {
        createOneUser(data: Any): User! upsertOneUser(where: Any create: Any update: Any): User! createManyUser(data: Any skipDuplicates: Any): AffectedRowsOutput! deleteOneUser(where: Any): User updateOneUser(data: Any where: Any): User updateManyUser(data: Any where: Any): AffectedRowsOutput! deleteManyUser(where: Any): AffectedRowsOutput! createOneTeam(data: Any): Team! upsertOneTeam(where: Any create: Any update: Any): Team! createManyTeam(data: Any skipDuplicates: Any): AffectedRowsOutput! deleteOneTeam(where: Any): Team updateOneTeam(data: Any where: Any): Team updateManyTeam(data: Any where: Any): AffectedRowsOutput! deleteManyTeam(where: Any): AffectedRowsOutput! createOneLeaderboardRow(data: Any): LeaderboardRow! upsertOneLeaderboardRow(where: Any create: Any update: Any): LeaderboardRow! createManyLeaderboardRow(data: Any skipDuplicates: Any): AffectedRowsOutput! deleteOneLeaderboardRow(where: Any): LeaderboardRow updateOneLeaderboardRow(data: Any where: Any): LeaderboardRow updateManyLeaderboardRow(data: Any where: Any): AffectedRowsOutput! deleteManyLeaderboardRow(where: Any): AffectedRowsOutput! executeRaw(query: Any parameters: Any): Json! queryRaw(query: Any parameters: Any): Json!
      }
      ,
      type AggregateUser {
        _count: UserCountAggregateOutputType _avg: UserAvgAggregateOutputType _sum: UserSumAggregateOutputType _min: UserMinAggregateOutputType _max: UserMaxAggregateOutputType
      }
      ,
      type UserGroupByOutputType {
        id: Int! email: String! name: String teamId: String createdAt: DateTime! updatedAt: DateTime! _count: UserCountAggregateOutputType _avg: UserAvgAggregateOutputType _sum: UserSumAggregateOutputType _min: UserMinAggregateOutputType _max: UserMaxAggregateOutputType
      }
      ,
      type AggregateTeam {
        _count: TeamCountAggregateOutputType _min: TeamMinAggregateOutputType _max: TeamMaxAggregateOutputType
      }
      ,
      type TeamGroupByOutputType {
        id: String! name: String! language: [Language] createdAt: DateTime! updatedAt: DateTime! _count: TeamCountAggregateOutputType _min: TeamMinAggregateOutputType _max: TeamMaxAggregateOutputType
      }
      ,
      type AggregateLeaderboardRow {
        _count: LeaderboardRowCountAggregateOutputType _avg: LeaderboardRowAvgAggregateOutputType _sum: LeaderboardRowSumAggregateOutputType _min: LeaderboardRowMinAggregateOutputType _max: LeaderboardRowMaxAggregateOutputType
      }
      ,
      type LeaderboardRowGroupByOutputType {
        leaderboardId: Int! userId: Int! rating: Int _count: LeaderboardRowCountAggregateOutputType _avg: LeaderboardRowAvgAggregateOutputType _sum: LeaderboardRowSumAggregateOutputType _min: LeaderboardRowMinAggregateOutputType _max: LeaderboardRowMaxAggregateOutputType
      }
      ,
      type AffectedRowsOutput {
        count: Int!
      }
      ,
      type UserCountOutputType {
        leaderboardRows: Int!
      }
      ,
      type UserCountAggregateOutputType {
        id: Int! email: Int! name: Int! teamId: Int! createdAt: Int! updatedAt: Int! _all: Int!
      }
      ,
      type UserAvgAggregateOutputType {
        id: Float
      }
      ,
      type UserSumAggregateOutputType {
        id: Int
      }
      ,
      type UserMinAggregateOutputType {
        id: Int email: String name: String teamId: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type UserMaxAggregateOutputType {
        id: Int email: String name: String teamId: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type TeamCountOutputType {
        users: Int!
      }
      ,
      type TeamCountAggregateOutputType {
        id: Int! name: Int! language: Int! createdAt: Int! updatedAt: Int! _all: Int!
      }
      ,
      type TeamMinAggregateOutputType {
        id: String name: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type TeamMaxAggregateOutputType {
        id: String name: String createdAt: DateTime updatedAt: DateTime
      }
      ,
      type LeaderboardRowCountAggregateOutputType {
        leaderboardId: Int! userId: Int! rating: Int! _all: Int!
      }
      ,
      type LeaderboardRowAvgAggregateOutputType {
        leaderboardId: Float userId: Float rating: Float
      }
      ,
      type LeaderboardRowSumAggregateOutputType {
        leaderboardId: Int userId: Int rating: Int
      }
      ,
      type LeaderboardRowMinAggregateOutputType {
        leaderboardId: Int userId: Int rating: Int
      }
      ,
      type LeaderboardRowMaxAggregateOutputType {
        leaderboardId: Int userId: Int rating: Int
      }
      ,
      type User {
        id: Int! email: String! name: String teamId: String leaderboardRows(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [LeaderboardRow] createdAt: DateTime! updatedAt: DateTime! Team: Team _count: UserCountOutputType!
      }
      ,
      type Team {
        id: String! name: String! language: [Language] users(where: Any orderBy: Any cursor: Any take: Any skip: Any distinct: Any): [User] createdAt: DateTime! updatedAt: DateTime! _count: TeamCountOutputType!
      }
      ,
      type LeaderboardRow {
        leaderboardId: Int! userId: Int! rating: Int User: User!
      }
      
  "
`;
